### ISSUE TITLE
Add : end-of-month and end-of-year calculation functionality for services and recurring invoices

### ISSUE THREAD

**Issue by lmag** [Dolibarr/dolibarr#32060]
### Feature Request

Add function of 

- [ ] Last day of month
- [ ] Last day of year
- [ ] dol_time_plus_duree should be in interface for reccuring invoice @W1W1-M @mdeweerd 

Instead of add number day of month like
![image](https://github.com/user-attachments/assets/cf8f24c3-b827-4069-9045-ff41e5f1a0d5)

**Comment by apio-sys** [comment]
Thanks for this fix, but it doesn't work as intended. See example below:

<img width="474" height="242" alt="Image" src="https://github.com/user-attachments/assets/dff639bf-824d-40fb-94e2-917b67aa9053" />

If the last generating was 30/09/2025, the next one should be on 31/10/2025. The only workaround at the moment is to set the generation data as 28th of the month and then come back each real last day of the month to correct the invoice.

@eldy Can you re-open the issue please or do you prefer I open a new one?

**Comment by eldy** [comment]
Currently, if we set a recurring date of 1 month and date is 30 septembre, next date should be 30 november because we want "every month".
Having a recurring date that is "the last day of month" is a different feature than "every month" so i reopen the ticket to get more info if this feature is implemented or not and how we can reproduce it.

**Comment by apio-sys** [comment]
Not really. Currently, if you set ie. end-of-month today 30/09, the next generation will be scheduled for 30/10 (not 30/11). So that means same day number in the next month hence will give problems when the number is higher than 28. The object of this ticket and PR is "Last day of month" so it should be 31/10 (and then 30/11 and so on). Actually, in `core/lib/date.lib.php` a function for that exists:

```
/**
 * Return GMT time for last day of a month or year.
 * Note: The timestamp contains last day and last hours (23:59:59)
 *
 *      @param          int                     $year           Year
 *      @param          int                     $month          Month
 *      @param          bool|int<0,1>|'gmt'|'tzserver'|'tzref'|'tzuser'|'tzuserrel'     $>
 *                                                                                  True >
 *      @return         int|''                                  Date as a timestamp, '' i>
 */
function dol_get_last_day($year, $month = 12, $gm = false)
{
        if ($year > 9999) {
                return '';
        }
        if ($month == 12) {
                $month = 1;
                $year += 1;
        } else {
                $month += 1;
        }

        // On se deplace au debut du mois suivant, et on retire un jour
        $datelim = dol_mktime(23, 59, 59, $month, 1, $year, $gm);
        $datelim -= (3600 * 24);

        return $datelim;
}
```

and it would make sense to use that here to make sure the last day of the month is selected as this feature request proposes but that isn't solved with this PR. Hence my bug report here #35561.

**Comment by mdeweerd** [comment]
1.
> Currently, if we set a recurring date of 1 month and date is 30 septembre, next date should be 30 november because we want "every month".

I am pretty sure that this is not what a user expects and this might even need to result in a n error/at least a warning ("This day does not exist in every month, some months may be skipped").

For this behavior there should be an option "every month with same day in month".

2. IMHO an acceptable approach would be that if the previous date was at the end of the month or > 28, then the new date is also at the end of the month.

3. There could be an option "every month, same delay to end of month".

So if the 30th, in a 31 day month, then it would be the 29th in a 30 month and the 27th in a 28 day month.

3. `dol_get_last_day()`
Luckilly there is no DST at the end of a month - finding the first day of the month at '00:00:00' and substracting one second would look more secure.

**Comment by apio-sys** [comment]
I have suggested following PR #36579 to "solve" this. At least it works as expected for me like this. I always get my invoice on the last day of the month if the day is higher than 28. And I think this is what most are looking for here. If this is acceptable and can be merged we could/should do the same for the recurring supplier invoices that suffer from the same issue.
